File: Courses/PHP Web Basics - май 2019/16.ExerciseEncapsulationInheritance/02.03.Mankind/Worker.php
<?php

require_once "Human.php";

class Worker extends Human
{
    /**
     * @var float
     */
    private $weekSalary;
    /**
     * @var float
     */
    private $hoursPerDay;

    public function __construct(string $firstName, string $lastName, float $weekSalary, float $hoursPerDay)
    {
        parent::__construct($firstName, $lastName);
        $this->setWeekSalary($weekSalary);
        $this->setHoursPerDay($hoursPerDay);
    }

    /**
     * @return float
     */
    public function getWeekSalary(): float
    {
        return $this->weekSalary;
    }

    /**
     * @param float $weekSalary
     */
    public function setWeekSalary(float $weekSalary): void
    {
        if ($weekSalary < 10) {
            throw new \Exception("Expected value mismatch!Argument: weekSalary");
        }
        $this->weekSalary = $weekSalary;
    }

    /**
     * @return float
     */
    public function getHoursPerDay(): float
    {
        return $this->hoursPerDay;
    }

    /**
     * @param float $hoursPerDay
     * @throws Exception
     */
    public function setHoursPerDay(float $hoursPerDay): void
    {
        if ($hoursPerDay < 1 || $hoursPerDay > 12) {
            throw new \Exception("Expected value mismatch!Argument: workHoursPerDay");
        }

        $this->hoursPerDay = $hoursPerDay;
    }
    const dayOfWeek = 7;
    /**
     * @return float
     */
    public function getSalaryPerHour() : float{
        //TO DO. Maybe we need setter, field and constructor?
        return number_format($this->getWeekSalary() / ($this->getHoursPerDay() * self::dayOfWeek), 2,'.', '');
    }

    public function __toString()
    {
        return "First Name: " . $this->getFirstName().PHP_EOL."Last Name: " . $this->getLastName().PHP_EOL.
            "Week Salary: " . number_format($this->getWeekSalary(), 2, '.','').PHP_EOL
            ."Hours per day: " . number_format($this->getHoursPerDay(),2, '.','').PHP_EOL
            ."Salary per hour: " . $this->getSalaryPerHour().PHP_EOL;

//        First Name: {worker's first name}
//        Last Name: {worker's second name}
//        Week Salary: {worker's salary}
//        Hours per day: {worker's working hours}
//        Salary per hour: {worker's salary per hour}
    }
}

File: Courses/C# OOP/10. Inheritance Exercise/Exercises_Inheritance/03.Mankind/Worker.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace _03.Mankind
{
    public class Worker : Human
    {
        private double weekSalary;
        private double workHoursPerDay;
        public Worker(string firstName, string lastName, double weekSalary, double workHoursPerDay) : base(firstName, lastName)
        {
            this.WeekSalary = weekSalary;
            this.WorkHoursPerDay = workHoursPerDay;
        }

        public double WeekSalary
        {
            get => weekSalary; 
            set
            {
                if (value <= 10)
                {
                    throw new ArgumentException("Expected value mismatch! Argument: weekSalary");
                }
                weekSalary = value;
            }
        }
        public double WorkHoursPerDay
        {
            get => workHoursPerDay; 
            set
            {
                if (value < 1 || value > 12)
                {
                    throw new ArgumentException("Expected value mismatch! Argument: workHoursPerDay");
                }
                workHoursPerDay = value;
            }
        }

        private double GetSalaryPerHour()
        {
            return WeekSalary / (WorkHoursPerDay * 5.0);
        }

        public override string ToString()
        {
            var sb = new StringBuilder();
            sb.AppendLine($"First Name: {FirstName}");
            sb.AppendLine($"Last Name: {LastName}");
            sb.AppendLine($"Week Salary: {WeekSalary:F2}");
            sb.AppendLine($"Hours per day: {WorkHoursPerDay:F2}");
            sb.AppendLine($"Salary per hour: {GetSalaryPerHour():F2}");
            return sb.ToString();
        }
    }
}
